### PR TITLE
Fix filteration logic for eval results

### LINKED ISSUE #10540: `trainer.test` prints duplicate results when `self.log(add_dataloader_idx=False)`
### To Reproduce

```py
import os

import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer, LightningDataModule


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len

class BoringData(LightningDataModule):
    def __init__(self):
        super().__init__()
    
    def test_dataloader(self):
        return [DataLoader(RandomDataset(32, 100), batch_size=16), DataLoader(RandomDataset(32, 100), batch_size=16)]


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)
    
    def test_step(self, batch, batch_idx, dataloader_idx):
        logits = self.layer(batch)
        loss = torch.nn.functional.mse_loss(logits, torch.randn(*logits.shape))
        self.log(f'test_loss_class_{dataloader_idx}', loss, add_dataloader_idx=False)


model = BoringModel()
dm = BoringData()
trainer = Trainer(fast_dev_run=True)
trainer.test(model, datamodule=dm)
```

Prints:
```console
[{'test_loss_class_0': 1.8550611734390259,
  'test_loss_class_1': 1.141597867012024},
 {'test_loss_class_0': 1.8550611734390259,
  'test_loss_class_1': 1.141597867012024}]
```

Expected:
```console
[{'test_loss_class_0': 1.8550611734390259},
 {'test_loss_class_1': 1.141597867012024}]
```

### Environment

- PyTorch Lightning Version (e.g., 1.3.0): master
- PyTorch Version (e.g., 1.8):
- Python version: 3.9
- OS (e.g., Linux): mac
- CUDA/cuDNN version:
- GPU models and configuration:
- How you installed PyTorch (`conda`, `pip`, source):
- If compiling from source, the output of `torch.__config__.show()`:
- Any other relevant information:

### Additional context

cc @tchaton @carmocca @edward-io @ananthsub

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #10540

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @carmocca @edward-io @ananthsub

### CODE REVIEW

File: tests/trainer/logging_/test_eval_loop_logging.py [hunk 1]
```diff
@@ -676,32 +675,6 @@ def val_dataloader(self):
...
-        ),
-        (
-            {"dl_idx": 3, "metrics": {"top_3_acc/dataloader_idx_0": 123, "top_3_acc/dataloader_idx_3": 321}},
-            {"top_3_acc/dataloader_idx_3": 321},
-        ),
-        # theoretical case, as `/dataloader_idx_3` would have been added
-        ({"dl_idx": 3, "metrics": {"top_3_acc": 123}}, {"top_3_acc": 123}),
-    ],
-)
-def test_filter_metrics_for_dataloader(kwargs, expected):
```
Comment (carmocca) [inline tests/trainer/logging_/test_eval_loop_logging.py:699]: rip unit test 😢

File: tests/trainer/logging_/test_eval_loop_logging.py [hunk 2]
```diff
@@ -676,32 +675,6 @@ def val_dataloader(self):
     trainer.fit(model)
 
 
-@pytest.mark.parametrize(
-    ["kwargs", "expected"],
-    [
```
Comment (carmocca) [inline tests/trainer/logging_/test_eval_loop_logging.py:681]: These cases should still be tested

File: pytorch_lightning/trainer/connectors/logger_connector/result.py [hunk 3]
```diff
@@ -527,14 +530,15 @@ def valid_items(self) -> Generator:
     def _forked_name(self, result_metric: ResultMetric, on_step: bool) -> Tuple[str, str]:
         name = result_metric.meta.name
         forked_name = result_metric.meta.forked_name(on_step)
+        add_datalaoder_idx = result_metric.meta.add_dataloader_idx
         dl_idx = result_metric.meta.dataloader_idx
-        if dl_idx is not None:
+        if add_datalaoder_idx and dl_idx is not None:
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/logger_connector/result.py:535]: This is an implementation detail from the loops leaking into the resultcollection, who is the `dl_idx is not None` check enough?

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 4]
```diff
@@ -155,21 +155,19 @@ def update_eval_step_metrics(self) -> None:
...
         num_dataloaders = self.trainer._evaluation_loop.num_dataloaders
         has_been_initialized = len(self.eval_loop_results) == num_dataloaders
-        for dl_idx in range(self.trainer._evaluation_loop.num_dataloaders):
-            # remove callback metrics that don't belong to this dataloader
-            callback_metrics = self._filter_metrics_for_dataloader(dl_idx, metrics)
+        assert self.trainer._results is not None
+        for dl_idx in range(num_dataloaders):
+            callback_metrics = self.trainer._results.metrics(
+                on_step, dataloader_idx=dl_idx if num_dataloaders > 1 else None
+            )["callback"]
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:169]:
```suggestion
            metrics = self.trainer._results.metrics(
                on_step, dataloader_idx=dl_idx if num_dataloaders > 1 else None
            )
            callback_metrics = metrics["callback"]
```

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 5]
```diff
@@ -155,21 +155,19 @@ def update_eval_step_metrics(self) -> None:
...
         if self.trainer.sanity_checking:
             return
 
+        on_step = not self._epoch_end_reached
         num_dataloaders = self.trainer._evaluation_loop.num_dataloaders
         has_been_initialized = len(self.eval_loop_results) == num_dataloaders
-        for dl_idx in range(self.trainer._evaluation_loop.num_dataloaders):
-            # remove callback metrics that don't belong to this dataloader
-            callback_metrics = self._filter_metrics_for_dataloader(dl_idx, metrics)
+        assert self.trainer._results is not None
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:165]: This should access the evaluation loop metrics directly, as we do for num dataloaders

**Review by carmocca — CHANGES_REQUESTED** [review]

File: pytorch_lightning/trainer/connectors/logger_connector/result.py [hunk 6]
```diff
@@ -527,14 +530,15 @@ def valid_items(self) -> Generator:
     def _forked_name(self, result_metric: ResultMetric, on_step: bool) -> Tuple[str, str]:
         name = result_metric.meta.name
         forked_name = result_metric.meta.forked_name(on_step)
+        add_datalaoder_idx = result_metric.meta.add_dataloader_idx
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/logger_connector/result.py:533]:
```suggestion
        add_dataloader_idx = result_metric.meta.add_dataloader_idx
```

File: pytorch_lightning/trainer/connectors/logger_connector/result.py (hunk 3 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/logger_connector/result.py:535]: we still pass in correct dataloader_idx regardless of `add_dataloader_idx. See the example in the linked issue.

File: tests/trainer/logging_/test_eval_loop_logging.py (hunk 2 above)
Comment (rohitgr7) [reply tests/trainer/logging_/test_eval_loop_logging.py:681]: have updated the current test with these cases.

**Review by tchaton — APPROVED** [review]
LGTM !

File: tests/trainer/logging_/test_eval_loop_logging.py (hunk 1 above)
Comment (awaelchli) [reply tests/trainer/logging_/test_eval_loop_logging.py:699]: is it not possible anymore to unit test this part?

**Review by awaelchli — APPROVED** [review]

File: tests/trainer/logging_/test_eval_loop_logging.py (hunk 1 above)
Comment (rohitgr7) [reply tests/trainer/logging_/test_eval_loop_logging.py:699]: @awaelchli unfortunately not. tbh I tried, but there are a lot of components involved here so unit testing isn't reliable here.

File: pytorch_lightning/trainer/connectors/logger_connector/result.py [hunk 7]
```diff
@@ -564,7 +568,9 @@ def any_tensor(_: Any) -> None:
                 metrics["log"][forked_name] = value
 
             # populate callback metrics. callback metrics don't take `_step` forked metrics
-            if self.training or result_metric.meta.on_epoch and not on_step:
+            if (self.training or result_metric.meta.on_epoch and not on_step) and (
+                dataloader_idx is None or result_metric.meta.dataloader_idx == dataloader_idx
+            ):
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/logger_connector/result.py:573]: I don't think this should be here, instead it should be

```py
    def valid_items(self, dataloader_idx: Optional[int] = None) -> Generator:
        """This function is used to iterate over current valid metrics."""
        return (
            (k, v)
            for k, v in self.items()
            if not (isinstance(v, ResultMetric) and v.has_reset)
            and (dataloader_idx in (None, v.meta.dataloader_idx))
        )
```

where this function calls

```py
for _, result_metric in self.valid_items(dataloader_idx=dataloader_idx):
```

Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/logger_connector/result.py:573]: will update!

File: pytorch_lightning/trainer/connectors/logger_connector/result.py [hunk 8]
```diff
@@ -449,7 +451,7 @@ def log(
         # storage key
         key = f"{fx}.{name}"
         # add dataloader_suffix to both key and fx
-        if dataloader_idx is not None:
+        if add_dataloader_idx and dataloader_idx is not None:
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/logger_connector/result.py:454]: I don't think these are just for internal separation. Try with`add_dataloader_idx=False` on master. `dataloader_idx` was set to None when `add_dataloader_idx=False` which means no suffix was added in that case, but now the value for `dataloader_idx` is independent of `add_dataloader_idx` here we need to check for it and add the suffix only when it's required.

**Review by carmocca — APPROVED** [review]
